### PR TITLE
Bind CommandPalette to Tabs and MRU Tabs collections

### PR DESCRIPTION
**Pull request by Don-Vito**
Second step towards https://github.com/microsoft/terminal/issues/8415.
Closes: #8374

Introduces:
* Biding of CommandPalette to Tabs and MRU tabs. Simplifies the terminal page and improves performance.
* Fix possible race when focusing by index
* Improves readability by introducing _GetFocusedTabImpl

### CODE REVIEW

**Comment by zadjii-msft** [comment]
Hey do you mind if I hold off reviewing this PR till after #8420 merges? It's just a little hard to tease out what's new in this PR the way it's currently set up 😬

**Comment by Don-Vito** [comment]
> Hey do you mind if I hold off reviewing this PR till after #8420 merges? It's just a little hard to tease out what's new in this PR the way it's currently set up 😬 

@zadjii-msft - absolutely! I can create a PR for the diff between the branches, but the idea is in any case to merge them one by one.

**Comment by DHowett** [comment]
QQ: This is going to be unmarked draft when it is retargeted to your merged work in main, right? :smile:

**Comment by Don-Vito** [comment]
> 
> 
> QQ: This is going to be unmarked draft when it is retargeted to your merged work in main, right? 😄

@DHowett - yes. Absolutely. I need to merge and improve few things. Hope to get to it on the weekend (to get some more focus time) :blush:

**Comment by Don-Vito** [comment]
@DHowett - btw.. not sure what is the best approach here.. but if the merge of the previous phase wasn't squashed the merge/rebase could be much simpler and safer.

**Comment by DHowett** [comment]
Yeah, our squash workflow can be a pain for stacked branches. A merge _should_ work, though.

Worst case, a rebase?

**Comment by Don-Vito** [comment]
> 
> 
> Yeah, our squash workflow can be a pain for stacked branches. A merge _should_ work, though.
> 
> Worst case, a rebase?

It will work.. no worries - I was just surprised with the amount of conflicts 😄

**Comment by Don-Vito** [comment]
@DHowett, @zadjii-msft - this was meant to be a minimal step of binding the tabs manually (later on they will be bound to filterable list control). However during the testing I met few bugs.. and it ended up more code than required (it fixes the race from #8374).
I can try to split this into several PRs.. but it won't be fun 😢

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalTab.cpp:233]: bare TODO

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2814]: Shouldn't we keep the `const`?

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2865]:
```suggestion
    // - tab: the tab to bump.
```

**Review by carlos-zamora — APPROVED** [review]

**Comment by Don-Vito** [comment]
@carlos-zamora  - thanks for the review! Due to https://github.com/microsoft/terminal/pull/8653 it seems that I need to add some more logic here. So I guess that at this point the right thing to do is to split this PR into several stacked ones:
* One for introducing `GetFocusedTabImpl`
* One for maximizing the usage of tabs
* One for binding tabs and ensuring (still don't know how that the titles and the icons keep getting updated

**Pull request closed without merging**
